src/connector/codec/src/decoder/avro/schema.rs

```diff
@@ -280,3 +314,54 @@ fn supported_avro_to_json_type(schema: &Schema) -> bool {
         | Schema::Union(_) => false,
     }
 }
+
+/// The field name when converting Avro union type to RisingWave struct type.
+pub(super) fn avro_schema_to_struct_field_name(schema: &Schema) -> String {
+    match schema {
+        Schema::Null => unreachable!(),
+        Schema::Union(_) => unreachable!(),
+        // Primitive types
+        Schema::Boolean => "boolean".to_string(),
+        Schema::Int => "int".to_string(),
+        Schema::Long => "long".to_string(),
+        Schema::Float => "float".to_string(),
+        Schema::Double => "double".to_string(),
+        Schema::Bytes => "bytes".to_string(),
+        Schema::String => "string".to_string(),
+        // Unnamed Complex types
+        Schema::Array(_) => "array".to_string(),
+        Schema::Map(_) => "map".to_string(),
+        // Named Complex types
+        // TODO: Verify is the namespace correct here
+        Schema::Enum(_) | Schema::Ref { name: _ } | Schema::Fixed(_) => todo!(),
+        Schema::Record(_) => schema.name().unwrap().fullname(None),
+        // Logical types
+        // XXX: should we use the real type or the logical type as the field name?
+        // It seems not to matter much, as we always have the index of the field when we get a Union Value.
+        //
+        // Currently choose the logical type because it might be more user-friendly.
+        //
+        // Example about the representation:
+        // schema: ["null", {"type":"string","logicalType":"uuid"}]
+        // data: {"string": "67e55044-10b1-426f-9247-bb680e5fe0c8"}
+        //
+        // Note: for union with logical type AND the real type, e.g., ["string", {"type":"string","logicalType":"uuid"}]
+        // In this case, the uuid cannot be constructed.
+        // Actually this should be an invalid schema according to the spec. https://issues.apache.org/jira/browse/AVRO-2380
+        // But some library like Python and Rust both allow it. See `risingwave_connector_codec::decoder::avro::tests::test_avro_lib_union`
```

Comment:
How about being conservative and reject unions containing logical types in the first version, given their behavior is inconsistent across different language implementations?

(As mention in the linked issue, logical types are allowed to be ignored, and it shall not make a difference for interpretability. The support within python & rust shall be considered as an unintentional bug.)

Reply:
> How about being conservative and reject unions containing logical types in the first version

Understand and I also used to be conservative, want to delay the decision. Especially since the codebase of our source contains too many mysterious stuff, being extra cautious can prevent us from more footguns..

But @fuyufjh seems to support more the "just do it" way, like the "map to jsonb" problem #16963:
- There seem to be no obvious problems of the decision. It might not be always the "best" way in all situations, but also not too bad (at most 1 or 2 operations away?).
- These usages are common, so "_I would be unwilling to say we support it (if these cases not supported)_". (Well this might be arguable, since why few users have requested union/map type before? 🤣)
- Very few additional workload is required.

I think these also make sense. 

Currently I tend to "just do it", if
- there's no serious drawbacks we can come up with
- the decision is reversible

Reply:
> the decision is reversible

No, it would become burden for backward compatibility, both for union-containing-logical-type and for map-containing-more-types. If we believe it is okay to break them in the future, meaning few people are relying on them, why implement it at the first place; if we foresee they would be used in this way, then we would need to support both flavors forever. @fuyufjh

Reply:
As we are the talking about the Avro ***Source***, I think we can be less restrictive, because the data's producer is not under the control of us (or even users).

Saying, if the user is using a strict Java Avro writer, then `["double", {"type":"int", "logicalType": "date"}]` won't exist. No problem.

Only when the user is using a non-strict Avro writer, we will be facing this problem. In that case, I think accepting the data is a bit better than rejecting it.

Reply:
> `struct<"double" double, "date" date>` vs `struct<"double" double, "int" date>`. They are not compatible and cannot co-exist. A decision has to be made and irreversible.

By "reversible" I mean introducing config option for compatibility (or just breaking change). Compared with large architecture decisions, this are "relatively reversible" to me. lol

> If we believe it is okay to break them in the future, meaning few people are relying on them, why implement it at the first place.

This is indeed a good question and perhaps I don't have a good answer to it. But I feel a lot of what we do is _guess what users want_?

> Again I am also leaning towards the former option but would like to defer the decision until we have more inputs on how it is being used in related systems.

What I doubt now is that whether we can have useful inputs from users? Similar to the "map to jsonb" problem, I guess if we ask users "what expected result do you want", the answer is probably "whatever"... Perhaps users can hardly understand the subtleties, and need "it just works". 

So for such kind of problems, we still finally have to choose for users. That's why I'd make the decision earlier.

Reply:
> Saying, if the user is using a strict Java Avro writer, then ["double", {"type":"int", "logicalType": "date"}] won't exist. No problem.

`["int", {"type":"int", "logicalType": "date"}]` won't exist. No problem. But `["double", {"type":"int", "logicalType": "date"}]` can exist, and its JSON encoding is `{"int": 19912}` rather than `{"date":"2024-07-08"}`.

> Only when the user is using a non-strict Avro writer, we will be facing this problem. In that case, I think accepting the data is a bit better than rejecting it.

We can separate the PR and put it aside. As soon as we learn such valid use case, it could be merged easily.

Reply:
Well, I think there are 2 problems mixed together:
1. The invalid schema: `["int", {"type":"int", "logicalType": "date"}]`. I guess @fuyufjh is talking about allowing it, because we want to be tolerate. This point makes some sense (like allowing out-of-range JSON numbers?), but currently Rust Avro lib will reject it.
2. Which field name to use for logical types (logical or actual).

Reply:
> `struct<"double" double, "int" date>`

BTW, I just realized that `"int" date` is way too strange.🤣

Reply:
> if we ask users "what expected result do you want", the answer is probably "whatever"... Perhaps users can hardly understand the subtleties, and need "it just works".

This is not always the case. Such answer is still informative and we would know whether our proposed solution fits their concrete needs, rather than *guess what users want*. It is true in this case there is still risk that decision is biased toward single use case, but better than for speculated zero use case.

Just thought about a new issue: Logical types may be based on named types (fixed or record), for example `[{"type":"fixed","name":"Decimal128","size":16,"logicalType":"decimal","precision":38,"scale":2}, {"type":"fixed","name":"Decimal256","size":32,"logicalType":"decimal","precision":50,"scale":2}]`. This is valid in Java (different name) but we may encounter a conflict due to use of `decimal` as field name. Given Avro spec has been vague (or against the idea) here, making our extension shall be careful.

Reply:
> BTW, I just realized that `"int" date` is way too strange. 🤣

Yes so I have been saying I also prefer logical type name, but need more input to justify. As my latest example shown, at least we shall do the following instead:
* When physical type is named, use physical `name`; else use logical type

^ I am hesitating because there may be more loopholes that need patches like this, and changing from `"decimal" decimal` to `"Decimal" decimal` is definitely a breaking change - of course we could introduce a new option. It just feels unnecessary complex to maintain these backward compatible options.

Reply:
I reconsidered why I'm thinking "make reversible decision early" or "bias for action" is good. Basically there are 2 main benefits:
1. To better serve potential users. Since users have limited attention, they may not bother telling us their requirements, but will just silently go away when they meet problems. So we'd better make everything smooth if we can already foresee the requirement.
2. (More importantly), unblock us to do later work. 

For the specific problem of source (more specifically, `FORMAT`s), it's actually the "**leaf decision**" (no follow-ups to be done...). Therefore, the point 2 doesn't hold.

The point 1 still holds to some extent to me. But since we are to-B software (so serious users, especially customers will not walk away silently just for this minor reason), and 
in the first place, few users requested these features like Avro Union and Map, this might be less likely a concern (but an imaginary problem). 

So now I would agree with you that making the decision early doesn't have outstanding benefits but potential risks, and splitting the seemingly controversial part into a separate PR is indeed more responsible. cc @neverchanje @fuyufjh